core/src/main/scala/org/apache/spark/sql/CachedDataFrame.scala

```diff
@@ -88,7 +88,9 @@ class CachedDataFrame(session: SparkSession, queryExecution: QueryExecution,
   private lazy val boundEnc = exprEnc.resolveAndBind(logicalPlan.output,
     sparkSession.sessionState.analyzer)
 
-  def queryExecutionString: String = queryExecution.toString()
+  private lazy val queryExecutionString: String = queryExecution.toString()
+
+  private lazy val numPartitions: Int = cachedRDD.getNumPartitions
```

Comment:
Won't this cause trouble if the pruning is different for different tokens? From what I know Soubhik had handled that by explicitly clearing the cached partitions inside cachedRDD and re-calculating in every execution.